### ISSUE TITLE
Generalise QC check that ensures that all terms fall under preferred root

### ISSUE THREAD

**Issue by matentzn** [INCATools/ontology-development-kit#358]
If there is a preferred root declaration, then all terms in the base-namespace must fall under at least one class with a preferred root tag. See https://github.com/obophenotype/cell-ontology/pull/725#issuecomment-700316778

**Comment by gouttegd** [comment]
@matentzn Do you want to add that to the `odk:normalize` command already used for #1175?

**Comment by matentzn** [comment]
Yes. i think as for the general pattern - we can try the feature in the ODK robot plugin, and if we think its useful, migrate it to ROBOT core?

**Comment by gouttegd** [comment]
> we can try the feature in the ODK robot plugin, and if we think its useful, migrate it to ROBOT core?

Sounds good.

I think this should be the general pattern for ROBOT plugins:

### (1) Is the command highly specific to one project?

Then make your own plugin. It’s easy, and we don’t want ROBOT to end up with a command for every single task imaginable (that was the mistake of OWLTools).

Typical examples would be Uberon’s `uberon:merge-species` and `uberon:create-species-subset` commands.

### (2) Can the command be useful for a handful of projects?

Then maybe consider making a distinct plugin, that is not tied to any single project.

### (3) Can the command be useful for _many_ projects?

Then consider adding it to the ODK plugin – especially if you think that command should be used as part of any standard ODK workflow.

Once the command had been in the ODK plugin for a while (let’s say at least a full ODK major release cycle), consider submitting it to upstream ROBOT.

### (4) Is the command intended to work around a bug in upstream ROBOT?

Then *submit a proper fix to ROBOT* first! And then only add the command to the ODK plugin (or any other plugin) as a *temporary* workaround that should be removed as soon as a version of ROBOT is available in which the bug is fixed.

Typical example would the current’s `odk:normalize --merge-axioms true`, which I hope to remove if/when https://github.com/ontodev/robot/pull/1240 is merged (or https://github.com/ontodev/robot/issues/1239 is fixed by any other means).

Likewise if the command implements a new feature that would fit nicely into one of the existing ROBOT commands (like `odk:normalize --add-source`, which I hope is a temporary workaround until https://github.com/ontodev/robot/pull/1242 is merged).

**Comment by matentzn** [comment]
I agree, this is a great way to put it!

**Comment by gouttegd** [comment]
@matentzn How do you envision that check to interact with the COB alignment check?

I can imagine three possibilities:

### (A) Completely independent checks

Checking that all terms fall the declared preferred root(s) (which I would henceforth call “self-alignment”) would be implemented in the same `odk:validate` command as COB alignment, but would work completely independently.

That is, `odk:validate` would have two modes:

* if an upper ontology is specified with the `--upper-ontology` (or `--upper-ontology-iri`) option, then the command checks alignment against the terms of that upper ontology;
* otherwise (no upper ontology specified), the command checks self-alignment.

To check both alignment against COB _and_ alignment against preferred roots, just repeat the command, as in:

```sh
$ robot odk:validate --input my-ontology.owl --report-output self-align-violations.tsv \
        odk:validate --upper-ontology-iri http://purl.obolibrary.org/obo/cob.owl --report-output cob-align-violations
```

I think this would be the most useful and flexible (useful because flexible) approach.

### (B) Check for alignment against _either_ COB roots _or_ preferred root(s)

The command behaves as if the preferred roots declared by the ontology were part of the upper ontology. That is, the ontology is considered “aligned” if all its classes are subclasses of any of the COB root class _or_ any of the preferred roots.

This would allow any ontology to “circumvent” COB by using its own roots. Whether it is a good thing or not is an open question (I am inclined to think it would _not_ be a good thing).

### (C) Check for alignment against COB roots _and_ preferred root(s)

The command checks that all classes are subclasses of a preferred root, _and_ that the preferred roots themselves are subclasses of a COB roots.

In effect, quite similar to (A), but less flexible.

**Comment by matentzn** [comment]
> That is, odk:validate would have two modes:

Can we maybe call this command `validate-alignment` to reduce ambiguity?

> otherwise (no upper ontology specified), the command checks self-alignment.

Self alignment done using IAO:000700?

> I think this would be the most useful and flexible (useful because flexible) approach.

Agreed. I like your idea! Awesome!

**Comment by gouttegd** [comment]
> Can we maybe call this command validate-alignment to reduce ambiguity?

It’s called `check-align` for now: https://incatools.github.io/odk-robot-plugin/alignment.html

> > otherwise (no upper ontology specified), the command checks self-alignment.
>
> Self alignment done using IAO:000700?

Yes. But in the end the behaviour implemented in the `check-align` command is that checking for “self-alignment” has to be explicitly requested (using `--use-self true`) – it is _not_ a default behaviour to fallback to in the absence of an upper ontology, contrary to what I initially proposed above.

**Comment by matentzn** [comment]
> It’s called check-align for now: https://incatools.github.io/odk-robot-plugin/alignment.html

Awesome. Thats best.

> Yes. But in the end the behaviour implemented in the check-align command is that checking for “self-alignment” has to be explicitly requested (using --use-self true) – it is not a default behaviour to fallback to in the absence of an upper ontology, contrary to what I initially proposed above.

Like it.

In this case we are totally in agreement.

There is a bit of something to be sorted out wrt alignment depth, but that can go in the future.

For example, as per current spec, you could align your entire ontology with https://www.ebi.ac.uk/ols4/ontologies/cob/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FBFO_0000040?lang=en. That I guess would be a legal alignment, but not a desired one;

Since this cannot be solved in all generality, I would think that the report is critical:

1. number of aligned classes per "upper level class"
2. perhaps counting not total subclasses but total subclasses where this root / upper level term is the most specific one.

That way you could tell that ok, you have 5000 classes aligned with `BFO:0000040`, none of which are aligned with any of its children.

Not sure either way. I need to find time to try the tools and see what would be useful...

**Comment by gouttegd** [comment]
> Since this cannot be solved in all generality

Could be if COB itself explicitly declared which of the classes it contains are “COB roots“ (with IAO:0000700).

Then you could (1) explicitly merge COB with the ontology you want to align, and (2) perform a “self-alignment” check:

```sh
robot merge -i cob.owl -i my-ontology.owl \
      odk:check-align --use-self true
```

Alternatively, without IAO:0000700 annotations in COB, you could explicitly list the COB roots:

```sh
robot merge -i cob.owl -i my-ontology.owl \
      odk:check-align --term COB:x --term COB:y ...
```

(But that has the obvious drawback that you need to maintain a list of COB roots somewhere; it would be much better if COB did that itself.)

**Comment by gouttegd** [comment]
> Could be if COB itself explicitly declared which of the classes it contains are “COB roots“ (with IAO:0000700).

I actually find astonishing that COB does _not_ do that.

How on Earth are people supposed to know what the COB roots are? I had assumed that all “first-level” classes in COB (all classes directly under `owl:Thing`) were COB roots by definition, but that is apparently not the case if BFO:0000040 is not considered a “COB root”.

**Comment by gouttegd** [comment]
As for the COB alignment issue: the basic feature is here. If you want a more detailed alignment report, please open a new ticket.
